backend/cmd/server.go

```diff
@@ -86,6 +92,70 @@ func main() {
 	})
 }
 
+// CacheMiddleWare is Middleware for Caching purpose.
+func CacheMiddleWare(c *HeadlampConfig) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			ctx, span := telemetry.CreateSpan(ctx, r, "cluster-api", "handleClusterAPI",
+				attribute.String("cluster", mux.Vars(r)["clusterName"]),
+			)
+
+			defer span.End()
+
+			contextKey, err := c.getContextKeyForRequest(r)
+			if err != nil {
+				c.handleError(w, ctx, span, err, "failed to get context key", http.StatusBadRequest)
+				return
+			}
+
+			kContext, err := c.KubeConfigStore.GetContext(contextKey)
+			if err != nil {
+				c.handleError(w, ctx, span, err, "failed to get context", http.StatusNotFound)
+				return
+			}
+
+			if kContext.Error != "" {
+				c.handleError(w, ctx, span, errors.New(kContext.Error), "context has error", http.StatusBadRequest)
+				return
+			}
+
+			rcw := k8cache.Initialize(w)
+
+			key, err := k8cache.GenerateKey(r.URL, contextKey)
+			if err != nil {
+				c.handleError(w, ctx, span, err, "failed to generate key ", http.StatusBadRequest)
+				return
+			}
+
+			isAllowed, err := k8cache.IsAllowed(r.URL, kContext, w, r)
+			if err != nil {
+				next.ServeHTTP(w, r)
+				return
+			}
```

Comment:
If the request is forwarded due to error the response for the request is not cached here for subsequent requests.